File: webapp/src/blocks/quickAction.test.ts
// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {createQuickAction, QuickAction} from './quickAction'

describe('quickAction block', () => {
    describe('createQuickAction', () => {
        test('should create a quick action with default values', () => {
            const action = createQuickAction()

            expect(action.id).toBe('')
            expect(action.name).toBe('')
            expect(action.style).toEqual({color: 'propColorDefault'})
            expect(action.confirmRequired).toBe(false)
            expect(action.confirmText).toBe('')
            expect(action.conditions).toEqual([])
            expect(action.actions).toEqual([])
        })

        test('should create a quick action with provided values', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
                style: {color: '#0066cc'},
                confirmRequired: true,
                confirmText: 'Are you sure?',
                conditions: [
                    {
                        propertyId: 'status-prop-id',
                        operator: 'in',
                        values: ['done'],
                    },
                ],
                actions: [
                    {
                        type: 'setProperty',
                        propertyId: 'status-prop-id',
                        value: 'in-progress',
                    },
                ],
            }

            const action = createQuickAction(partialAction)

            expect(action.id).toBe('test-id')
            expect(action.name).toBe('Test Action')
            expect(action.style).toEqual({color: '#0066cc'})
            expect(action.confirmRequired).toBe(true)
            expect(action.confirmText).toBe('Are you sure?')
            expect(action.conditions).toHaveLength(1)
            expect(action.conditions[0]).toEqual({
                propertyId: 'status-prop-id',
                operator: 'in',
                values: ['done'],
            })
            expect(action.actions).toHaveLength(1)
            expect(action.actions[0]).toEqual({
                type: 'setProperty',
                propertyId: 'status-prop-id',
                value: 'in-progress',
            })
        })

        test('should merge partial values with defaults', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
            }

            const action = createQuickAction(partialAction)

            expect(action.id).toBe('test-id')
            expect(action.name).toBe('Test Action')
            expect(action.style).toEqual({color: 'propColorDefault'})
            expect(action.confirmRequired).toBe(false)
            expect(action.confirmText).toBe('')
            expect(action.conditions).toEqual([])
            expect(action.actions).toEqual([])
        })

        test('should handle confirmRequired being explicitly set to false', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
                confirmRequired: false,
            }

            const action = createQuickAction(partialAction)

            expect(action.confirmRequired).toBe(false)
        })

        test('should handle confirmRequired being explicitly set to true', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
                confirmRequired: true,
                confirmText: 'Confirm?',
            }

            const action = createQuickAction(partialAction)

            expect(action.confirmRequired).toBe(true)
            expect(action.confirmText).toBe('Confirm?')
        })

        test('should create quick action with multiple conditions', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
                conditions: [
                    {
                        propertyId: 'status-prop-id',
                        operator: 'in',
                        values: ['waiting', 'todo'],
                    },
                    {
                        propertyId: 'priority-prop-id',
                        operator: 'in',
                        values: ['high', 'critical'],
                    },
                ],
            }

            const action = createQuickAction(partialAction)

            expect(action.conditions).toHaveLength(2)
            expect(action.conditions[0].propertyId).toBe('status-prop-id')
            expect(action.conditions[1].propertyId).toBe('priority-prop-id')
        })

        test('should create quick action with multiple actions', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
                actions: [
                    {
                        type: 'setProperty',
                        propertyId: 'status-prop-id',
                        value: 'in-progress',
                    },
                    {
                        type: 'addComment',
                        text: 'Work started',
                    },
                ],
            }

            const action = createQuickAction(partialAction)

            expect(action.actions).toHaveLength(2)
            expect(action.actions[0].type).toBe('setProperty')
            expect(action.actions[1].type).toBe('addComment')
        })

        test('should create quick action with all supported operators', () => {
            const operators = [
                'in',
                'not in',
                'empty',
                'not empty',
                '>',
                '<',
                '>=',
                '<=',
                'equal',
                'contains',
                'not contains',
                'checked',
                'not checked',
            ] as const

            operators.forEach((op) => {
                const partialAction: Partial<QuickAction> = {
                    id: `test-${op}`,
                    name: `Test ${op}`,
                    conditions: [
                        {
                            propertyId: 'prop-id',
                            operator: op,
                            values: op === 'empty' || op === 'not empty' || op === 'checked' || op === 'not checked' ? [] : ['value'],
                        },
                    ],
                }

                const action = createQuickAction(partialAction)

                expect(action.conditions[0].operator).toBe(op)
            })
        })

        test('should create quick action with all supported action types', () => {
            const actionTypes = ['setProperty', 'clearProperty', 'addComment'] as const

            actionTypes.forEach((actionType) => {
                const partialAction: Partial<QuickAction> = {
                    id: `test-${actionType}`,
                    name: `Test ${actionType}`,
                    actions: [
                        {
                            type: actionType,
                            propertyId: actionType === 'addComment' ? undefined : 'prop-id',
                            value: actionType === 'setProperty' ? 'value' : undefined,
                            text: actionType === 'addComment' ? 'comment text' : undefined,
                        },
                    ],
                }

                const action = createQuickAction(partialAction)

                expect(action.actions[0].type).toBe(actionType)
            })
        })

        test('should create quick action with {current_user} special value', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Assign to Current User',
                actions: [
                    {
                        type: 'setProperty',
                        propertyId: 'assignee-prop-id',
                        value: '{current_user}',
                    },
                ],
            }

            const action = createQuickAction(partialAction)

            expect(action.actions[0].value).toBe('{current_user}')
        })

        test('should create quick action with {now} special value', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Set to Now',
                actions: [
                    {
                        type: 'setProperty',
                        propertyId: 'date-prop-id',
                        value: '{now}',
                    },
                ],
            }

            const action = createQuickAction(partialAction)

            expect(action.actions[0].value).toBe('{now}')
        })

        test('should handle empty conditions array', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
                conditions: [],
            }

            const action = createQuickAction(partialAction)

            expect(action.conditions).toEqual([])
        })

        test('should handle empty actions array', () => {
            const partialAction: Partial<QuickAction> = {
                id: 'test-id',
                name: 'Test Action',
                actions: [],
            }

            const action = createQuickAction(partialAction)

            expect(action.actions).toEqual([])
        })
    })
})

File: server/model/quickAction_test.go
// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package model

import (
	"testing"
)

func TestQuickAction_IsValid(t *testing.T) {
	tests := []struct {
		name      string
		action    QuickAction
		expectErr bool
		errMsg    string
	}{
		{
			name: "Valid quick action with setProperty",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Start Work",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions: []QuickActionCondition{
					{
						PropertyID: "status-prop-id",
						Operator:   QuickActionOpIn,
						Values:     []string{"waiting-id"},
					},
				},
				Actions: []QuickActionAction{
					{
						Type:       QuickActionSetProperty,
						PropertyID: "status-prop-id",
						Value:      "in-progress-id",
					},
				},
			},
			expectErr: false,
		},
		{
			name: "Valid quick action with addComment",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Add Comment",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions:      []QuickActionCondition{},
				Actions: []QuickActionAction{
					{
						Type: QuickActionAddComment,
						Text: "Work started",
					},
				},
			},
			expectErr: false,
		},
		{
			name: "Valid quick action with confirmation",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Delete Card",
				Style: map[string]string{
					"color": "#cc0000",
				},
				ConfirmRequired: true,
				ConfirmText:     "Are you sure you want to delete this card?",
				Conditions:      []QuickActionCondition{},
				Actions: []QuickActionAction{
					{
						Type:       QuickActionSetProperty,
						PropertyID: "status-prop-id",
						Value:      "deleted-id",
					},
				},
			},
			expectErr: false,
		},
		{
			name: "Valid quick action with clearProperty",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Clear Assignee",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions:      []QuickActionCondition{},
				Actions: []QuickActionAction{
					{
						Type:       QuickActionClearProperty,
						PropertyID: "assignee-prop-id",
					},
				},
			},
			expectErr: false,
		},
		{
			name: "Valid quick action with multiple actions",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Start Work and Comment",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions:      []QuickActionCondition{},
				Actions: []QuickActionAction{
					{
						Type:       QuickActionSetProperty,
						PropertyID: "status-prop-id",
						Value:      "in-progress-id",
					},
					{
						Type: QuickActionAddComment,
						Text: "Work started",
					},
				},
			},
			expectErr: false,
		},
		{
			name: "Invalid quick action with empty ID",
			action: QuickAction{
				ID:   "",
				Name: "Test Action",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions:      []QuickActionCondition{},
				Actions: []QuickActionAction{
					{
						Type: QuickActionAddComment,
						Text: "Test",
					},
				},
			},
			expectErr: true,
			errMsg:    "quick action ID is required",
		},
		{
			name: "Invalid quick action with empty name",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions:      []QuickActionCondition{},
				Actions: []QuickActionAction{
					{
						Type: QuickActionAddComment,
						Text: "Test",
					},
				},
			},
			expectErr: true,
			errMsg:    "quick action name is required",
		},
		{
			name: "Invalid quick action with confirmRequired but empty confirmText",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Test Action",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: true,
				ConfirmText:     "",
				Conditions:      []QuickActionCondition{},
				Actions: []QuickActionAction{
					{
						Type: QuickActionAddComment,
						Text: "Test",
					},
				},
			},
			expectErr: true,
			errMsg:    "confirmation text is required when confirmation is enabled",
		},
		{
			name: "Invalid quick action with no actions",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Test Action",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions:      []QuickActionCondition{},
				Actions:         []QuickActionAction{},
			},
			expectErr: true,
			errMsg:    "quick action must have at least one action",
		},
		{
			name: "Invalid quick action with condition missing propertyId",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Test Action",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions: []QuickActionCondition{
					{
						PropertyID: "",
						Operator:   QuickActionOpIn,
						Values:     []string{"value"},
					},
				},
				Actions: []QuickActionAction{
					{
						Type: QuickActionAddComment,
						Text: "Test",
					},
				},
			},
			expectErr: true,
			errMsg:    "condition propertyId is required",
		},
		{
			name: "Invalid quick action with invalid operator",
			action: QuickAction{
				ID:   "test-action-id",
				Name: "Test Action",
				Style: map[string]string{
					"color": "#0066cc",
				},
				ConfirmRequired: false,
				ConfirmText:     "",
				Conditions: []QuickActionCondition{
					{
						PropertyID: "prop-id",
						Operator:   QuickActionConditionOperator("invalid"),
						Values:     []string{"value"},
					},
				},
				Actions: []QuickActionAction{
					{
						Type: QuickActionAddComment,
						Text: "Test",
					},
				},
			},
			expectErr: true,
			errMsg:    "invalid operator",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := tt.action.IsValid()
			if (err != nil) != tt.expectErr {
				t.Errorf("expected error: %v, got: %v", tt.expectErr, err)
			}
			if err != nil && tt.expectErr && tt.errMsg != "" {
				if err.Error() != tt.errMsg && !containsSubstring(err.Error(), tt.errMsg) {
					t.Errorf("expected error message to contain %q, got: %q", tt.errMsg, err.Error())
				}
			}
		})
	}
}

func TestQuickActionAction_IsValid(t *testing.T) {
	tests := []struct {
		name      string
		action    QuickActionAction
		expectErr bool
		errMsg    string
	}{
		{
			name: "Valid setProperty action",
			action: QuickActionAction{
				Type:       QuickActionSetProperty,
				PropertyID: "status-prop-id",
				Value:      "in-progress-id",
			},
			expectErr: false,
		},
		{
			name: "Valid clearProperty action",
			action: QuickActionAction{
				Type:       QuickActionClearProperty,
				PropertyID: "assignee-prop-id",
			},
			expectErr: false,
		},
		{
			name: "Valid addComment action",
			action: QuickActionAction{
				Type: QuickActionAddComment,
				Text: "This is a comment",
			},
			expectErr: false,
		},
		{
			name: "Valid setProperty with {current_user} special value",
			action: QuickActionAction{
				Type:       QuickActionSetProperty,
				PropertyID: "assignee-prop-id",
				Value:      "{current_user}",
			},
			expectErr: false,
		},
		{
			name: "Valid setProperty with {now} special value",
			action: QuickActionAction{
				Type:       QuickActionSetProperty,
				PropertyID: "due-date-prop-id",
				Value:      "{now}",
			},
			expectErr: false,
		},
		{
			name: "Invalid setProperty with missing propertyId",
			action: QuickActionAction{
				Type:  QuickActionSetProperty,
				Value: "value",
			},
			expectErr: true,
			errMsg:    "propertyId is required for setProperty action",
		},
		{
			name: "Invalid clearProperty with missing propertyId",
			action: QuickActionAction{
				Type: QuickActionClearProperty,
			},
			expectErr: true,
			errMsg:    "propertyId is required for clearProperty action",
		},
		{
			name: "Invalid addComment with empty text",
			action: QuickActionAction{
				Type: QuickActionAddComment,
				Text: "",
			},
			expectErr: true,
			errMsg:    "text is required for addComment action",
		},
		{
			name: "Invalid action type",
			action: QuickActionAction{
				Type: QuickActionActionType("invalid"),
			},
			expectErr: true,
			errMsg:    "invalid action type",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := tt.action.IsValid()
			if (err != nil) != tt.expectErr {
				t.Errorf("expected error: %v, got: %v", tt.expectErr, err)
			}
			if err != nil && tt.expectErr && tt.errMsg != "" {
				if err.Error() != tt.errMsg && !containsSubstring(err.Error(), tt.errMsg) {
					t.Errorf("expected error message to contain %q, got: %q", tt.errMsg, err.Error())
				}
			}
		})
	}
}

func TestIsValidQuickActionOperator(t *testing.T) {
	tests := []struct {
		name     string
		operator QuickActionConditionOperator
		expected bool
	}{
		{"in operator", QuickActionOpIn, true},
		{"not in operator", QuickActionOpNotIn, true},
		{"empty operator", QuickActionOpEmpty, true},
		{"not empty operator", QuickActionOpNotEmpty, true},
		{"> operator", QuickActionOpGreaterThan, true},
		{"< operator", QuickActionOpLessThan, true},
		{">= operator", QuickActionOpGreaterEqual, true},
		{"<= operator", QuickActionOpLessEqual, true},
		{"equal operator", QuickActionOpEqual, true},
		{"contains operator", QuickActionOpContains, true},
		{"not contains operator", QuickActionOpNotContains, true},
		{"checked operator", QuickActionOpChecked, true},
		{"not checked operator", QuickActionOpNotChecked, true},
		{"invalid operator", QuickActionConditionOperator("invalid"), false},
		{"empty operator string", QuickActionConditionOperator(""), false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := isValidQuickActionOperator(tt.operator)
			if result != tt.expected {
				t.Errorf("expected %v, got %v", tt.expected, result)
			}
		})
	}
}

func containsSubstring(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 ||
		(len(s) > len(substr) && containsStringHelper(s, substr)))
}

func containsStringHelper(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}
